src/nccl_ofi_rdma.c

```diff
@@ -3818,24 +3835,68 @@ static int recv_close_deferred(nccl_net_ofi_recv_comm_t *recv_comm)
 	return ret;
 }
 
+static int alloc_rdma_flush_req(nccl_net_ofi_rdma_recv_comm_t *r_comm,
+					void *buff, size_t size,
+					nccl_net_ofi_rdma_mr_handle_t *buff_mr_handle,
+					nccl_net_ofi_rdma_req_t **ret_req)
+{
+	nccl_net_ofi_rdma_ep_t *ep = (nccl_net_ofi_rdma_ep_t *)r_comm->base.base.ep;
+	nccl_net_ofi_rdma_device_t *device = (nccl_net_ofi_rdma_device_t *)get_device_from_ep(ep);
+	nccl_net_ofi_scheduler_t *scheduler = device->scheduler;
+	int dev_id = r_comm->base.base.dev_id;
+	rdma_req_flush_data_t *flush_data = NULL;
+	*ret_req = NULL;
+
+	/* Allocate NCCL OFI request */
+	nccl_net_ofi_rdma_req_t *req = allocate_req(r_comm->nccl_ofi_reqs_fl);
+	if (OFI_UNLIKELY(req == NULL)) {
+		NCCL_OFI_WARN("Unable to get NCCL OFI request for device %d",
+			      dev_id);
+		return -ENOMEM;
+	}
+	req->comm = &r_comm->base.base;
+	req->dev_id = dev_id;
+	req->type = NCCL_OFI_RDMA_FLUSH;
+	req->free = free_flush_req;
+
+	flush_data = get_flush_data(req);
+	flush_data->data = buff;
+	flush_data->mr_handle = buff_mr_handle;
+	flush_data->total_num_compls = 0;
+
+	for (int rail_seq = 0; rail_seq < ep->num_rails; rail_seq++) {
+		flush_data->schedule[rail_seq] = scheduler->get_schedule(scheduler, size, ep->num_rails);
```

Comment:
I'm not sure I would use the scheduler here at all. You want to make sure you get exactly one transfer on each rail. The scheduler round-robins, so if another thread tried to grab a schedule in the middle, you might end up not getting four unique rails here.

Reply:
Also, it's confusing that `rail_seq != schedule->rail_id`.

Reply:
@rauteric yep, scheduler is handy, but you are right. get_schedule() itself has a locking `scheduler->rr_lock` inside, but during this for-loop, other thread can be scheduled. So, we either need a special mode of scheduler logic, which stride into all rails even for small size (4-byte), or we just do not use scheduler here.

regarding `rail_seq`, that's intentional because rail_id is implicitly mapped to actual NIC (EFA VF) ID in other places. However, it is the iterating sequence number through all rails in this for-loop.